proto/prost/private/protoc_wrapper.rs

```diff
@@ -0,0 +1,567 @@
+                .expect("Failed to split on '.'")
+                .1
+                .to_string();
+        }
+
+        module_info.insert(
+            module_name.clone(),
+            Module {
+                name,
+                contents: fs::read_to_string(path).map_err(from_error)?,
+                submodules: BTreeSet::new(),
+            },
+        );
+
+        let module_parts = module_name.split('.').collect::<Vec<&str>>();
+        for parent_module_index in 0..module_parts.len() {
+            let child_module_index = parent_module_index + 1;
+            if child_module_index >= module_parts.len() {
+                break;
+            }
+            let full_parent_module_name = module_parts[0..parent_module_index + 1].join(".");
+            let parent_module_name = module_parts[parent_module_index];
+            let child_module_name = module_parts[child_module_index];
+
+            module_info
+                .entry(full_parent_module_name.clone())
+                .and_modify(|parent_module| {
+                    parent_module
+                        .submodules
+                        .insert(child_module_name.to_string());
+                })
+                .or_insert(Module {
+                    name: parent_module_name.to_string(),
+                    contents: "".to_string(),
+                    submodules: [child_module_name.to_string()].iter().cloned().collect(),
+                });
+        }
+    }
+
+    let mut content = "// @generated\n\n".to_string();
+    write_module(&mut content, &module_info, "", 0)?;
+    Ok(content)
+}
+
+fn write_module(
+    content: &mut String,
+    module_info: &BTreeMap<String, Module>,
+    module_name: &str,
+    depth: usize,
+) -> ProtocResult<()> {
+    if module_name.is_empty() {
+        for submodule_name in module_info.keys() {
+            write_module(content, module_info, submodule_name, depth + 1)?;
+        }
+        return Ok(());
+    }
+    let module = module_info.get(module_name).expect("Failed to get module");
+    let indent = "  ".repeat(depth);
+    let is_rust_module = module.name != "_";
+
+    if is_rust_module {
+        content
+            .write_str(&format!("{}pub mod {} {{\n", indent, module.name))
+            .map_err(from_error)?;
+    }
+
+    content.write_str(&module.contents).map_err(from_error)?;
+
+    for submodule_name in module.submodules.iter() {
+        write_module(
+            content,
+            module_info,
+            [module_name, submodule_name].join(".").as_str(),
+            depth + 1,
+        )?;
+    }
+
+    if is_rust_module {
+        content
+            .write_str(&format!("{}}}\n", indent))
+            .map_err(from_error)?;
+    }
+
+    Ok(())
+}
+
+/// Compute the `--extern_path` flags for a list of proto files. This is
+/// expected to convert proto files into a list of
+/// `.example.prost.helloworld=crate_name::example::prost::helloworld`
+fn compute_proto_package_info(
```

Comment:
Is this something you could add a unit test for? Same for other functions in this file

Reply:
In order to setup these tests appropriately, I would need to recreate the protoc compilation environment (or very near it). I think it would be a lot more valuable to ensure the tests under `//proto/prost/private/tests` thoroughly cover all of the use-cases. That will result in more legible test cases without a bunch of setup boilerplate.

Let me know what you think though.

Reply:
I think if you split out calls to protoc from this function you should be fine. Then you can have a unit test describe what the expected output is which would be helpful in tracking regressions. Does that sound reasonable?